core/agent/src/main/java/org/apache/spark/raydp/Agent.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.raydp;
+
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.lang.instrument.Instrumentation;
+import java.lang.management.ManagementFactory;
+
+
+public class Agent {
+
+  public static final PrintStream DEFAULT_ERR_PS = System.err;
+
+  public static final PrintStream DEFAULT_OUT_PS = System.out;
+
+  public static void premain(String agentArgs, Instrumentation inst)
+      throws IOException {
+    // redirect system output/error stream so that annoying SLF4J warnings
+    // and other logs during binding
+    // SLF4J factory don't show in spark-shell
+    // Instead, the warnings and logs are kept in
+    // <ray session>/logs/slf4j-<worker pid>.log
+
+    String pid = ManagementFactory.getRuntimeMXBean().getName()
+        .split("@")[0];
+    String logDir = System.getProperty("ray.logging.dir");
+    if (logDir == null) {
+      logDir = "/tmp/ray/process-" + pid;
+      System.getProperties().put("ray.logging.dir", logDir);
+    }
+
+    File parentDir = new File(logDir);
+    if (!parentDir.exists()) {
+      parentDir.mkdirs();
+    }
+
+    File logFile = new File(parentDir, "/slf4j-" + pid + ".log");
```

Comment:
It contains below logs for each executor and master. If I don't capture them, they'll be print in pyspark-shell.
"
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/home/jiafu/anaconda3/envs/ray/lib/python3.9/site-packages/raydp/jars/raydp-agent-1.6.0-SNAPSHOT.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/jiafu/anaconda3/envs/ray/lib/python3.9/site-packages/pyspark/jars/log4j-slf4j-impl-2.17.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/jiafu/anaconda3/envs/ray/lib/python3.9/site-packages/ray/jars/ray_dist.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
mapped factory class: org.apache.logging.slf4j.Log4jLoggerFactory. load org.apache.logging.slf4j.Log4jLoggerFactory from file:/home/jiafu/anaconda3/envs/ray/lib/python3.9/site-packages/pyspark/jars/log4j-slf4j-impl-2.17.2.jar
SLF4J: Actual binding is of type [org.apache.logging.slf4j.Log4jLoggerFactory]
"

Reply:
> Maybe we do not need a separate dir for every process, as we are naming the file with pid?

I think we need a separate file for each process. Otherwise, we get messed up logs from different processes. It's not good for later tracing.

Reply:
> I think we need a separate file for each process. Otherwise, we get messed up logs from different processes. It's not good for later tracing.

Yes, that's true. But they are in different dirs, right? As you created dir above also with process id. We can share the same dir, different filename is enough.

Reply:
I've changed parentDir to same dir for all processes as you suggested yesterday. Please check latest code.